javascript/modules/client/src/internal/encoding/client.ts

```diff
@@ -4,13 +4,139 @@ import { BigNumber } from "@ethersproject/bignumber";
 import { AddressZero } from "@ethersproject/constants";
 import { AVAILABLE_CLIENT_FUNCTION_SIGNATURES } from "../constants/encoding";
 import { FunctionFragment, Interface, Result } from "@ethersproject/abi";
-import { IWithdrawParams } from "../interfaces/client";
+import {
+  IFreezeParams,
+  IPermissionParams,
+  IWithdrawParams,
+  PermissionIds,
+} from "../interfaces/client";
+import { keccak256 } from "@ethersproject/keccak256";
+import { toUtf8Bytes } from "@ethersproject/strings";
 
 export function getFunctionFragment(data: Uint8Array): FunctionFragment {
   const hexBytes = bytesToHex(data, true);
   const inter = new Interface(AVAILABLE_CLIENT_FUNCTION_SIGNATURES);
   return inter.getFunction(hexBytes.substring(0, 10));
 }
+export function encodeFreezeAction(
+  params: IFreezeParams,
+): Uint8Array {
+  const daoInterface = DAO__factory.createInterface();
+  const args = unwrapFreezeParams(params);
+  // get hex bytes
+  const hexBytes = daoInterface.encodeFunctionData("freeze", args);
+  return hexToBytes(strip0x(hexBytes));
+}
+
+export function decodeFreezeActionData(data: Uint8Array): IFreezeParams {
+  const daoInterface = DAO__factory.createInterface();
+  const hexBytes = bytesToHex(data, true);
+  const receivedFunction = daoInterface.getFunction(
+    hexBytes.substring(0, 10) as any,
+  );
+  const expectedFunction = daoInterface.getFunction("freeze");
+  if (receivedFunction.name !== expectedFunction.name) {
+    throw new Error(
+      "The received action is different from the expected one",
+    );
+  }
+  const result = daoInterface.decodeFunctionData("freeze", data);
+  return wrapFreezeParams(result);
+}
+
+function unwrapFreezeParams(
+  params: IFreezeParams,
+): [string, string] {
+  return [
+    params.where,
+    keccak256(toUtf8Bytes(params.permission)),
+  ];
+}
+function wrapFreezeParams(
+  result: Result,
+): IFreezeParams {
+  return {
+    where: result[0],
+    permission:
+      // @ts-ignore
+      Object.keys(PermissionIds).find((k) => PermissionIds[k] === result[1])
+        .replace("_ID", "") || "",
```

Comment:
No, the error has nothing to do with this, it has to do with the permission constant
```
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'.ts(7053)
  ```

Reply:
Uhm, then we could replace `{}` by something similar to `{} as { [k: string]: Permissions[string] }`  ?

Reply:
`{} as { [k: string]: string }` did the trick, thank you :)